---
title: "Hydrothermal Vent Viruses Protocol"
author: "Maggie Langwig"
date: "8/21/2022"
output:
  html_document: rmdformats::downcute
  toc: yes
  toc_float: yes
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Organizing the 43 assemblies

## Renaming
1. 11 assemblies do not have the sample name associated with the file so I am adding it.
```{bash, eval = FALSE}
cd /storage1/data12/Projects/Vent_Viruses/Assemblies_Renamed

for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\.fasta/,x)}1' $i >> $i.renamed; done
```

2. Special naming for S009:
```{bash, eval = FALSE}
for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\_metaspades_scaffolds.min1000.fasta/,x)}1' $i >> $i.renamed; done
```

3. Adding site identifier to assembly names (after checking your sed first!)
### Brothers UC
```{bash, eval = FALSE}
sed -i 's/>/>Brothers_UC_/g' S010_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_UC_/g' S011_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_UC_/g' S147_metaspades_scaffolds.min1000.fasta
```
### Brothers LC
```{bash, eval = FALSE}
sed -i 's/>/>Brothers_LC_/g' S014_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_LC_/g' S016_metaspades_scaffolds.min1000.fasta
```
### Brothers NWCA
```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCA_/g' S013_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCA_/g' S017_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCA_/g' S142_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCA_/g' S143_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCA_/g' S144_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCA_/g' S145_metaspades_scaffolds.min1000.fasta
```
### Brothers NWCB
```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCB_/g' S012_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCB_/g' S139_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCB_/g' S140_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCB_/g' S141_metaspades_scaffolds.min1000.fasta
```

```{bash, eval = FALSE}
sed -i 's/>/>Brothers_NWCB_/g' S146_metaspades_scaffolds.min1000.fasta
```
### Brothers Diffuse
```{bash, eval = FALSE}
sed -i 's/>/>Brothers_Diffuse_/g' S009_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>Brothers_Diffuse_/g' S015_metaspades_scaffolds.min1000.fasta
```
### ELSC Abe
```{bash, eval = FALSE}
sed -i 's/>/>ELSC_Abe_/g' A1_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Abe_/g' A3_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Abe_/g' 128-326.fasta
```
### ELSC Tui Malila
```{bash, eval = FALSE}
sed -i 's/>/>ELSC_Tui_Malila_/g' T10_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Tui_Malila_/g' T11_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Tui_Malila_/g' T2_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Tui_Malila_/g' 132-544.fasta

sed -i 's/>/>ELSC_Tui_Malila_/g' 134-614.fasta
```
### ELSC Bowl
```{bash, eval = FALSE}
sed -i 's/>/>ELSC_Bowl_/g' M1_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Bowl_/g' M2_metaspades_scaffolds.min1000.fasta
```
### ELSC Mariner
```{bash, eval = FALSE}
sed -i 's/>/>ELSC_Mariner_/g' 131-447.fasta

sed -i 's/>/>ELSC_Mariner_/g' M10_metaspades_scaffolds.min1000.fasta

sed -i 's/>/>ELSC_Mariner_/g' M17_metaspades_scaffolds.min1000.fasta
```
### ELSC Vai Lili
```{bash, eval = FALSE}
sed -i 's/>/>ELSC_Vai_Lili_/g' V2_metaspades_scaffolds.min1000.fasta
```
### Guaymas
```{bash, eval = FALSE}
sed -i 's/>/>Guaymas_/g' 4559-240.fasta

sed -i 's/>/>Guaymas_/g' 4561-380.fasta

sed -i 's/>/>Guaymas_/g' 4561-384.fasta

sed -i 's/>/>Guaymas_/g' 4571-419.fasta
```
### EPR
```{bash, eval = FALSE}
sed -i 's/>/>EPR_/g' 4281-140.fasta

sed -i 's/>/>EPR_/g' PIR-30.fasta
```
### MAR Rainbow
```{bash, eval = FALSE}
sed -i 's/>/>MAR_Rainbow_/g' 354-166.fasta

sed -i 's/>/>MAR_Rainbow_/g' 355-202.fasta
```
### MAR Lucky
```{bash, eval = FALSE}
sed -i 's/>/>MAR_Lucky_/g' 356-284.fasta

sed -i 's/>/>MAR_Lucky_/g' 356-308.fasta
```

After scaffold renaming, I used `mv` commands to rename files.

# Identifying viruses using VIBRANT
VIBRANT filters to remove scaffolds <1kb so not filtering assembly before. ID's dsDNA, ssDNA, and RNA viruses
```{bash, eval = FALSE}
for i in Assemblies_Renamed/*.fasta; do VIBRANT_run.py -i $i -t 15; done
```

## Count the number of viruses identified by VIBRANT
```{bash, eval = FALSE}
grep ">" -c VIBRANT_*/VIBRANT_phages*/*combined.fna | cut -f2 -d : | paste -sd+ | bc
```
Total of 53,135 viruses identified

# Run CheckV
CheckV provides quality estimates of viral genomes. I'm running this first to determine how many medium and high confidence viruses we have. If it's a lot I'll only focus on these.
```{bash, eval = FALSE}
find -type f -name "*combined.fna" | xargs cp -t ../Virus_Genomes/fna/

nohup checkv end_to_end Virus_Genomes/fna/All_Viruses.fna CheckV_Output -t 20 &
```

Check how many medium and high quality viruses:
```{bash,  eval = FALSE}
grep -v 'Low-quality\|Not-determined' quality_summary.tsv | wc
```
Result: 1,165 medium and high quality viruses (exclude header)
146 complete viruses
1,031 proviruses, or integrated viruses

# Grab only the medium and high quality viruses
Grab only the medium and high quality viruses predicted by CheckV
```{bash,  eval = FALSE}
grep -v 'Low-quality\|Not-determined' quality_summary.tsv > MedHighQual_Viruses.tsv

cut -f1 MedHighQual_Viruses.tsv > MedHighQual_Viruses_list.txt 

perl /storage1/data12/scripts/screen_list_new.pl CheckV_Output/MedHighQual_Viruses_list.txt Virus_Genomes/fna/All_Viruses.fna keep > MedHighQual_Viruses.fna
```

Cp the faa viruses into a new folder

```{bash,  eval = FALSE}
find -type f -name "*combined.faa" | xargs cp -t ../Virus_Genomes/faa
```

Get a list of the faa scaffold names for medium and high quality viruses:
```{bash,  eval = FALSE}
for i in `cat ../../CheckV_Output/MedHighQual_Viruses_list.txt`; do grep $i *.faa >> MedHighQual_faa_list.txt; done

sed 's/>/\t/g' MedHighQual_faa_list.txt | cut -f2 > MedHighQual_faa_list_final.txt
```

Extract the faa seqs by name:
```{bash,  eval = FALSE}
perl /storage1/data12/scripts/screen_list_new.pl MedHighQual_faa_list_final.txt All_Viruses.faa keep > MedHighQual_Viruses.faa
```

# Organizing the MAGs
I want to associate the site names and file name of each MAG to the scaffold header. I am not being as careful about renaming for the MAGs as I was for the assemblies, meaning I am just using general site name and not specific (e.g. Brothers vs Brothers_UC).

## Brothers Volcano
```{bash,  eval = FALSE}
pwd

/storage1/data12/Projects/Vent_Viruses/MAGs_Renamed/Brothers_Volcano

rename 's/scaf2bin.//g' *.fasta

rename 's/^/Brothers_/g' *.fasta

for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\.fasta/,x)}1' $i >> $i.renamed; done
```

## ELSC
```{bash,  eval = FALSE}
rename 's/scaf2bin.//g' *.fasta

rename 's/^/ELSC_/g' *.fasta

for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\.fasta/,x)}1' $i >> $i.renamed; done
```

## EPR
```{bash,  eval = FALSE}
rename 's/scaf2bin.//g' *.fasta

rename 's/^/EPR_/g' *.fasta

for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\.fasta/,x)}1' $i >> $i.renamed; done
```

## Guaymas
```{bash,  eval = FALSE}
rename 's/scaf2bin.//g' *.fasta

rename 's/^/Guaymas_/g' *.fasta

for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\.fasta/,x)}1' $i >> $i.renamed; done
```

## MAR
```{bash,  eval = FALSE}
rename 's/scaf2bin.//g' *.fasta

rename 's/^/MAR_/g' *.fasta

for i in *.fasta; do awk '/>/{sub(">","&"FILENAME"_");sub(/\.fasta/,x)}1' $i >> $i.renamed; done
```

# Read Mapping

## EPR
```{bash,  eval = FALSE}
conda activate

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/EPR/EPR_4281-140.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/EPR/*.fastq.gz -o Read_Mapping/EPR -t 20 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/EPR/EPR_PIR-30.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/EPR/*.fastq.gz -o Read_Mapping/EPR -t 20 &
```

## Guaymas
```{bash,  eval = FALSE}
nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/Guaymas/Guaymas_4559-240.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/Guaymas/*.fastq.gz -o Read_Mapping/Guaymas -t 20 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/Guaymas/Guaymas_4561-380.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/Guaymas/*.fastq.gz -o Read_Mapping/Guaymas -t 20 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/Guaymas/Guaymas_4561-384.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/Guaymas/*.fastq.gz -o Read_Mapping/Guaymas -t 20 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/Guaymas/Guaymas_4571-419.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/Guaymas/*.fastq.gz -o Read_Mapping/Guaymas -t 20 &
```

## MAR

```{bash,  eval = FALSE}
nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/MAR_Lucky_356-284.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/MAR/*.fastq.gz -o Read_Mapping/MAR -t 20 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/MAR_Lucky_356-308.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/MAR/*.fastq.gz -o Read_Mapping/MAR -t 15 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/MAR_Rainbow_354-166.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/MAR/*.fastq.gz -o Read_Mapping/MAR -t 20 &

nohup python3 /storage1/data14/software/mapping/mapper.py -i Assemblies_Renamed/MAR_Rainbow_355-202.fasta -r /storage1/Reads/HydroVents_ZhouReysenbach/MAR/*.fastq.gz -o Read_Mapping/MAR -t 20 &
```

## Brothers and ELSC
Now trying with Cody's mapping wrapper
```{bash,  eval = FALSE}
conda activate

nohup python3 /storage1/data14/software/mapping/mapper.py -i AssembliesToReadsFinal.txt --batch -o Read_Mapping/BrothersELSC -rd /storage1/Reads/HydroVents_ZhouReysenbach -fd /storage1/data12/Projects/Vent_Viruses/Assemblies_Renamed -t 20 &
```

# Remove viral contamination from MAGs

BLAST all viruses against all the MAGs. You are looking for 100% matches and 100% coverage. Unless it is a lysogenic virus, it should be contamination. Just because a virus is binned with a MAG does not mean it belongs to that virus.

1. Run BLAST search
```{bash,  eval = FALSE}
cat *.fasta > All_VentMAGs.fna

makeblastdb -in All_VentMAGs.fna -title "Vent_MAGs" -dbtype nucl

nohup blastn -query  /storage1/data12/Projects/Vent_Viruses/Virus_Genomes/fna/All_Viruses.fna -db /storage1/data12/Projects/Vent_Viruses/MAGs_Renamed/Remove_ViralContam/All_VentMAGs.fna -out MAGs_Renamed/Remove_ViralContam/VirusHits_toMAGs.txt -num_threads 20 -outfmt "6 qseqid length qlen slen pident bitscore stitle" &
```

2. Filter blast hits and subset from MAGs
```{bash,  eval = FALSE}
python3 /storage1/data12/scripts/filter_blast_100c_pi.py VirusHits_toMAGs.txt VirusHits_toMAGs_filtered.txt

cut -f7 VirusHits_toMAGs_filtered.txt > Viral_Contam.txt

for i in *.fasta; do perl /storage1/data12/scripts/screen_list_new.pl Remove_ViralContam/Viral_Contam.txt $i >> Remove_ViralContam/$i.noVirusContam.fna; done

cat *noVirusContam* > All_VentMAGs_noViralContam.fna
```

# ViWrap
**Unfortunately this ended up not being an option because it would take a month to run through all my samples. One smaller scale sample took 10 hours, times 42 samples = inefficient.**

Testing ViWrap - it would be nice if I could use it to get all the output I need :)

## Brothers Diffuse
```{bash,  eval = FALSE}
conda activate /storage1/data22/ViWrap_conda_environments/ViWrap

ViWrap run --input_metagenome /storage1/data12/Projects/Vent_Viruses/Assemblies_Renamed/Brothers_Volcano/Brothers_Diffuse_S009_metaspades_scaffolds.min1000.fasta \
               --input_reads /storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane2-s009-indexN705-S505-GGACTCCT-CTCCTTAC-279_debarcode_1.fastq.gz,/storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane2-s009-indexN705-S505-GGACTCCT-CTCCTTAC-279_debarcode_2.fastq.gz,/storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane4-s015-indexN703-S517-AGGCAGAA-TCTTACGC-Green_debarcode_1.fastq.gz,/storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane4-s015-indexN703-S517-AGGCAGAA-TCTTACGC-Green_debarcode_2.fastq.gz \
               --input_reads_type illumina \
               --reads_mapping_identity_cutoff 0.97 \
               --out_dir /storage1/data12/Projects/Vent_Viruses/ViWrap/Brothers_Diffuse \
               --db_dir /storage1/data22/ViWrap/ViWrap_db \
               --identify_method  vb \
               --conda_env_dir /storage1/data22/ViWrap_conda_environments \
               --threads 20 \
               --input_length_limit 2000 \
               --custom_MAGs_dir /storage1/data12/Projects/Vent_Viruses/MAGs_Renamed/Remove_ViralContam/Brothers/Diffuse
               

nohup ./bash_BD.sh &
```

I decided to leave off the custom MAG option for iphop since that takes a lot of time. I can take all the viruses and all the MAGs and run that on my own after I have all the vMAGs and unbinned viruses.

```{bash,  eval = FALSE}
conda activate /storage1/data22/ViWrap_conda_environments/ViWrap

ViWrap run --input_metagenome /storage1/data12/Projects/Vent_Viruses/Assemblies_Renamed/Brothers_Volcano/Brothers_Diffuse_S015_metaspades_scaffolds.min1000.fasta \
               --input_reads /storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane2-s009-indexN705-S505-GGACTCCT-CTCCTTAC-279_debarcode_1.fastq.gz,/storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane2-s009-indexN705-S505-GGACTCCT-CTCCTTAC-279_debarcode_2.fastq.gz,/storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane4-s015-indexN703-S517-AGGCAGAA-TCTTACGC-Green_debarcode_1.fastq.gz,/storage1/Reads/HydroVents_ZhouReysenbach/Brothers_Volcano/Diffuse_Flow/lane4-s015-indexN703-S517-AGGCAGAA-TCTTACGC-Green_debarcode_2.fastq.gz \
               --input_reads_type illumina \
               --reads_mapping_identity_cutoff 0.97 \
               --out_dir /storage1/data12/Projects/Vent_Viruses/ViWrap/Brothers_Diffuse_S015 \
               --db_dir /storage1/data22/ViWrap/ViWrap_db \
               --identify_method  vb \
               --conda_env_dir /storage1/data22/ViWrap_conda_environments \
               --threads 20 \
               --input_length_limit 2000
               

nohup ./bash_BD_S015.sh &
```

## Running ViWrap using Cody's wrapper

```{bash,  eval = FALSE}
conda activate

nohup python3 /storage1/data14/for_maggie/ViWrap_loop.py -s AssembliesToReads_ViWrap.txt -a /storage1/data12/Projects/Vent_Viruses/Assemblies_Renamed/ -r /storage1/Reads/HydroVents_ZhouReysenbach/ &
```



